gearbot_2_lib/src/datastore/crypto.rs

```diff
@@ -0,0 +1,69 @@
+use std::borrow::Cow;
+use std::fmt;
+
+use aes_gcm::aead::generic_array::{typenum::U32, GenericArray};
+use aes_gcm::{
+    aead::{Aead, NewAead},
+    Aes256Gcm,
+};
+use rand::{thread_rng, RngCore};
+
+/// An encryption key that is used to secure guild data.
+pub struct EncryptionKey<'a>(Cow<'a, GenericArray<u8, U32>>);
+
+impl<'a> EncryptionKey<'a> {
+    /// Constructs a new `EncryptionKey` from some bytes,
+    /// returning a key tied to the raw values.
+    pub fn construct_borrowed(bytes: &'a [u8]) -> Self {
+        Self(Cow::Borrowed(GenericArray::from_slice(bytes)))
+    }
+
+    /// Constructs a new `EncryptionKey` from some bytes,
+    /// returning a key that is free of ownership requirements.
+    pub fn construct_owned(bytes: &[u8]) -> Self {
+        Self(Cow::Owned(GenericArray::clone_from_slice(bytes)))
+    }
+}
+
+impl fmt::Debug for EncryptionKey<'_> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.write_str("EncryptionKey")
+    }
+}
+
+pub fn generate_guild_encryption_key(main_encryption_key: &EncryptionKey, guild_id: u64) -> Vec<u8> {
+    let mut csprng = thread_rng();
+    let mut guild_encryption_key = [0u8; 32];
+    csprng.fill_bytes(&mut guild_encryption_key);
+
+    encrypt_bytes(&guild_encryption_key, main_encryption_key, guild_id)[..32].to_vec()
+}
+
+pub fn encrypt_bytes(plaintext: &[u8], key: &EncryptionKey, msg_id: u64) -> Vec<u8> {
+    let aead = Aes256Gcm::new(&key.0);
+
+    // Since nonce's only never need to be reused, and Discor's snowflakes for messages
```

Comment:
I do love reading my own questionable code and comments from almost a year ago 🤣

Reply:
let's consider this whole crypto file a placeholder for something more proper later. the byte truncation i ended up doing lower down isn't really a good solution to the problem either :P